SPEC.md

```diff
@@ -0,0 +1,225 @@
+  complete(){ console.error('request complete'); },
+});
+```
+
+A fetcher can delegate to other fetchers. For example a polling fetcher can send a query operation to an http fetcher on a specified interval. Each request is routed through that http fetcher and returned through the observable returned by the polling fetcher. Using this type of delegation, fetchers can form stacks that each add their own functionality.
+
+<p align="center">
+  <br>
+  <img src="images/polling-stack.png" alt="Http Fetcher"/>
+</p>
+
+
+```js
+const pollingFetcher = new PollingFetcher(
+  10000, //Polling Interval in ms
+  new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' })
+);
+```
+
+In addition to the query, `Operation` includes a context, which is modified and passed down the stack of fetchers. In addition, this context is sent to the server to through the query body. This context could be used by a polling fetcher to tell a caching fetcher a request should be returned from the cache or not.
+
+<p align="center">
+  <br>
+  <img src="images/context.png" alt="Fetcher Stack"/>
+</p>
+
+## API
+
+The fetcher interface contains a single method:
+
+```js
+interface ApolloFetcher{
+  request(Operation) : Observable<FetchResult>
+}
+```
+
+An Operation contains the AST for a particular query, variables and associated context:
+
+```js
+Operation {
+  query: DocumentNode,
+  operationName: string,
+  variables: object,
+  context: object,
+}
+```
+
+An `Observable` is returned from the `request` function and controls the life-cycle of a GraphQL request:
+
+```js
+Observable{
+  //Causes the underlying producer to start
+  subscribe(Subscriber<FetchResult>) : UnsubscribeHandler
+  subscribe(
+    next(FetcheResult) => void,
+    error(any) => void,
+    complete() => void) : UnsubscribeHandler
+  status(): any
+}
+```
+
+A Subscriber is passed to the `subscribe` function
+
+```js
+Subscriber<T>{
+  next(T),
+  error(any),
+  complete(),
+}
+```
+
+UnsubscribeHandler is returned from a `subscribe` function in the Observable
+
+```js
+type UnscubscribeHandler = () => void
+```
+
+FetchResult is passed to the `next` function in the Observable
+
+```js
+FetchResult{
+  data?: object,
+  errors?: object[],
+  extensions?: any,
+  context?:object,
+}
+```
+
+* Should explain what might go into context.
+
+## More examples
+
+Here is a list of planned fetchers with selected diagrams:
+
+Base
+
+* Websocket: subscriptions support
+* Batching-\*: support for batching operations with a transport
+
+<p align="center">
+  <br>
+  <img src="images/batch-fetcher.png" alt="Batch Fetcher"/>
+</p>
+
+Intermediate
+
+* Polling: repeats requests on a specified interval
+
+<p align="center">
+  <br>
+  <img src="images/polling-fetcher.png" alt="Polling Fetcher"/>
+</p>
+
+* Caching: returns data if result in the cache and stores data in cache on response
+* Compose: combines a list of fetchers (would require additional semantics around constructor argument order)
```

Comment:
I'm not sure what this one would do. How is it different from split? or is this like the higher-order `compose` function which would let you chain fetchers?

Writing that just made me realize that we're still missing a code example of how you would stack multiple fetchers. I think we should have something like a `compose` function, but we'll have to figure out the details. For now, maybe you can just make a code example of manually connecting them?